### DISCUSSION TITLE
Impact on making forall a keyword

### DISCUSSION THREAD

**Discussion by nomeata** [ghc-proposals/ghc-proposals#440] [General]
This discussion relates to proposal https://github.com/ghc-proposals/ghc-proposals/pull/281 which would, eventually, make `forall` a proper keyword everywhere (types and terms), and it would be a keyword irrespective of the set of enabled extensions.

This would break code that uses `forall` as an identifier. In particular, libraries that have a symbol called `forall` in their public API would have to change that API before they can use the latest version of GHC. A [quick search](https://hackage-search.serokell.io/?q=%5Eforall+%3A%3A) and then some manually selection (excluding those where `forall` only appears in tests) points to these libraries as most seriously affected. 

 * [HaTeX](https://hackage-search.serokell.io/viewfile/HaTeX-3.22.3.0/Text/LaTeX/Base/Math.hs#line-942)
 * [HasCacBDD](https://hackage-search.serokell.io/viewfile/HasCacBDD-0.1.0.3/hs/Data/HasCacBDD.hs#line-134)
 * [Hsmtlib](https://hackage-search.serokell.io/viewfile/Hsmtlib-2.8.8.8/Hsmtlib/HighLevel.hs#line-14)
 * [atp](https://hackage-search.serokell.io/viewfile/atp-0.1.0.0/src/ATP/FirstOrder/Smart.hs#line-278)
 * [boolector](https://hackage-search.serokell.io/viewfile/boolector-0.0.0.13/src/Boolector/Foreign.chs#line-734)
 * [constraints](https://hackage-search.serokell.io/viewfile/constraints-0.13/src/Data/Constraint/Forall.hs#line-196)
 * [copilot-language](https://hackage-search.serokell.io/viewfile/copilot-language-3.5/src/Copilot/Language/Spec.hs#line-185)
 * [Dhall](https://hackage-search.serokell.io/viewfile/dhall-1.40.1/src/Dhall/Diff.hs#line-126)
 * [ersatz](https://hackage.haskell.org/package/ersatz-0.4.9/docs/Ersatz-Variable.html#v:forall)
 * [expressions](https://hackage.haskell.org/package/expressions-0.5/docs/Data-Expression.html#v:forall)
 * [flowlocks-framework](https://hackage-search.serokell.io/viewfile/flowlocks-framework-0.1.3.1/src/Security/InfoFlow/Policy/FlowLocks/Lattice.hs#line-135)
 * [haskelzinc](https://hackage.haskell.org/package/haskelzinc-0.3.1.1/docs/Interfaces-MZAST.html#v:forall)
 * [haskhol-core](https://hackage-search.serokell.io/viewfile/haskhol-core-1.1.0/src/HaskHOL/Core/Lib.hs#line-924)
 * [idris](https://hackage-search.serokell.io/viewfile/idris-1.3.3/src/Idris/Core/Elaborate.hs#line-437)
 * [lmonad](https://hackage.haskell.org/package/lmonad-0.1.0.0/docs/LMonad-Label-DisjunctionCategory.html#v:forall)
 * [monadiccp](https://hackage.haskell.org/package/monadiccp-0.7.7/docs/Data-Expr-Sugar.html#v:forall)
 * [odbb](https://hackage.haskell.org/package/obdd-0.8.2/docs/OBDD-Operation.html#v:forall)
 * [quickcheck-state-machine](https://hackage.haskell.org/package/quickcheck-state-machine-0.7.1/docs/Test-StateMachine-Logic.html#v:forall)
 * [satchmo](https://hackage.haskell.org/package/satchmo-2.9.9.3/docs/Satchmo-Boolean.html#v:forall)
 * [sbv](https://hackage.haskell.org/package/sbv-8.16/docs/Data-SBV.html#v:forall)
 * [smtlib2](https://hackage.haskell.org/package/smtlib2-1.0/docs/Language-SMTLib2.html#v:forall)
 * [sparsecheck](https://hackage.haskell.org/package/sparsecheck-0.1.0.3/docs/SparseCheck.html#v:forall)
 * [tamarin-theorem-prover](https://hackage.haskell.org/package/tamarin-prover-theory-0.8.6.0/docs/Theory-Model-Formula.html#v:forall)
 * [what4](https://hackage.haskell.org/package/what4-1.2.1/docs/What4-Protocol-SMTLib2-Syntax.html#v:forall)
 
I _did_ mean to contact the authors of the affected packages, and invite them here to have a discussion on how bad the proposed change would be for them, and which migration strategy works best for them, but that was when I thought it’s 5 affected packages. Guess it’s more… so I am reluctant to now send out dozends of emails. WDYT, @goldfirere @int-index?

**Comment by int-index** [comment]
No problem, I’ll contact the maintainers of these packages. Thanks for putting the list together!

**Comment by vagarenko** [comment]
I randomly clicked on a few links on that list, looks like a lot of those packages also define `exists`. Were there any plans for making it a keyword too?

**Comment by m4lvin** [comment]
I am the author of HasCacBDD, which is probably one of the least used packages on that list. Personally I would not find it a big problem as I can find some other name or workaround in case `forall` becomes a keyword. In fact, for many other logical functions this is already necessary(*) and common, for example I used `con` (for conjunction) to not clash with `and` from the Prelude etc.

That said, there is a theme behind the list of packages: most of them are about (first order) Logic or for some other reason need a definitions of formulas which contain quantifiers. In such projects this use of `forall` to me seems an obvious and very common choice for a function name and it would be a pity to have to work around it. Besides projects on hackage there might also be other (non-public or just not on Hackage) software in the area of Logic and Model Checking that would break with this.

I assume `forAll` could be used as an alternative and would not be affected by the keyword?

(*) well, not really *necessary*, I guess `and` can be hidden as it is just a function and not a keyword.

**Reply by nomeata** [reply]
> I assume forAll could be used as an alternative and would not be affected by the keyword?

Correct

**Reply by jwaldmann** [reply]
> example I used `con` (for conjunction) to not clash with `and` from the Prelude

the solution for that problem is to use https://hackage.haskell.org/package/ersatz-0.4.9/docs/Ersatz-Bit.html#t:Boolean  always (even when accessing Prelude operations)?

I maintain https://hackage.haskell.org/package/obdd , and I provide BDD functions via that class - which, however, has no `forall`. I could  follow some re-naming agreement by logic API authors. But to me, both the suggested `forall_` and `forAll` look ugly, and I might prefer `for_all`.

Anyway, I think my package isn't widely used (by others, by me), so the renaming is only a minor annoyance - while progress in GHC (in this case, for writing types) is much more important. So by all means, do go ahead.

**Comment by Daniel-Diaz** [comment]
Hi! HaTeX maintainer here. I would probably change the `forall` to be `forall_`, which I already did for, for example,  `pi_`.

**Comment by Gabriella439** [comment]
I'm totally fine with this proposal and this would not cause me major issues.  I'd just rename identifiers named `forall` to `forall_` as @Daniel-Diaz suggested

**Comment by LeventErkok** [comment]
Speaking for `sbv`: Indeed this'd be a breaking change, but not a big deal. SBV user base is nimble enough to accommodate.

I'd like to voice the concern @vagarenko raised though. SBV also defines `exists`; which GHC might also want to reserve in the future. This might be a good time to do both to reduce future conflicts.

**Reply by Gabriella439** [reply]
I don't believe we should bundle a breaking change to `exists` with this, for the reasons outlined in this post: https://www.haskellforall.com/2019/05/release-early-and-often.html

**Reply by jwaldmann** [reply]
NB:  domain name checks out. you also have haskellexists.com?

**Comment by treeowl** [comment]
@Gabriel439, I think there's a bit more incentive in this case. Suppose someone uses both `forall` and `exists`. If GHC takes `forall`, they may change to `forAll`. Then if GHC later takes `exists`, they're a bit stuck; the capitalization convention doesn't apply. So either they need to use two conventions or change `forAll` to something else.

**Reply by Ericson2314** [reply]
Yeah I like `forAll` but that does kinda do it in. Oh well.

**Comment by treeowl** [comment]
One option would be to produce just a *warning* for `exists`, until GHC takes it or decides not to.

**Comment by vagarenko** [comment]
Just remembered, there is also `foreach` planned.

**Comment by aztek** [comment]
[atp](https://hackage.haskell.org/package/atp) is my small pet project. It won't be a big deal for me to rename `forall` to `forAll` to support a future version of GHC. Like others, my package also defines `exists`. A similar future renaming for it could be `forSome`, though it's less common in first-order logic.

**Comment by ekmett** [comment]
I'm happy to rename combinators in `constraints` and `ersatz` should the proposal pass.

**Comment by goldfirere** [comment]
Very good point about `exists`. I [unsurprisingly](https://richarde.dev/papers/2021/exists/exists.pdf) think we should reserve `exists` at the same time.

Thanks to all respondents so far for being so accommodating!

**Reply by phadej** [reply]
What about `foreach`? Would there be dependent, retained `exists` as well?

**Reply by int-index** [reply]
I’m not as worried about `foreach` and `exists` – they might stay niche for a while and it will be easy to hide them behind a language pragma.

The issue with `forall` is that `ExplicitForAll` is ubiquitous.

**Reply by phadej** [reply]
True, but not true. For example, `role` is supposedly hidden behind the language extensions, however it isn't: https://gitlab.haskell.org/ghc/ghc/-/issues/18941

I also suspect it will cause so implementation problems in pretty-printer. if we can have `exists` type variable, but then pretty-print types in a module which enables `exists`-extension, what will happen? Is it easy to do alpha-rename in pretty-printer? Can all cases be renamed?

**Comment by int-index** [comment]
Thanks all! The proposal has been accepted, `forall` will become a keyword.

**Comment by robdockins** [comment]
FWIW, the use of `forall` in What4 is fairly incidental, I think we'll probably migrate to `forall_`.

**Comment by AntC2** [comment]
BTW GHC is rather inconsistent as to its case treatment around `forall`.

* `-XExplicitForAll` -- that upper-`A` often trips me up.
* `-fprint-explicit-foralls, -Wunused-foralls` -- no upper `A`, but why plural?
